### ISSUE TITLE
Test Promregator in a HA setup

### ISSUE THREAD

**Issue by eaglerainbow** [promregator/promregator#44]
It should be possible to run Promregator in a high-availability (HA) setup:
* Two distinct physical servers each one run an own Promregator instance
* They share the same configuration (local copy of the configuration)
* a HA-reverse proxy dispatches requests to servers only up and running; on failure automatic fallback to the second instance is happening.

So far, this approach is entirely untested and thus undocumented. A manual test shall be conducted to verify, if Promregator works in this scenario properly.

**Comment by vervas** [comment]
If running multiple containers of promregator counts as such (which I believe it is) then I have some feedback.

Since promregator is stateless there is no issue scaling the service out. Also given that they all follow the same naming conventions for the targets, discovery, etc all of those instance yield the same results as if running one.

Things get a bit problematic when it comes to monitoring. Since these instances are load balanced, you get their metrics in an a rotating manner. This is expected since promregator does not tag it's own metrics, nor applies CF service discovery on itself (unless added as a target in `promregator.yaml` but then it should omit it's own metrics appended on it ;)). The other solution would be configure prometheus to scrape each instance as an individual target.

In both scenarios, it might make sense to consider putting the promregator metrics in a separate endpoint of their own.

**Comment by eaglerainbow** [comment]
> If running multiple containers of promregator counts as such (which I believe it is)

I'd say: yes.

> This is expected since promregator does not tag it's own metrics,

Would it improve the situation for you, if it did (e.g. adding the docker container's ID as tag)?

> nor applies CF service discovery on itself

Well, it kind of does 😁 with the endpoint of `/promregatorMetrics` (see also the [endpoint page](https://github.com/promregator/promregator/blob/ce2e6a269c8a4bafd8b288be1304f65a4878d73b/docs/endpoint.md)). This endpoint also gets added as target in the json document at `/discovery`.  But you are right in that sense, that it does not discover its neighbors. However, I am a little hesitant to do such a thing, as you don't know by definition where Promregator is being used (you might be running it as docker container plainly on a host, the docker image could be running in a CF cell, or you could run docker manually even without any docker involved...)

> In both scenarios, it might make sense to consider putting the promregator metrics in a separate endpoint of their own.

With the things said just above, I don't think that I am able to understand your statement properly. Could you please elaborate a little what you are missing comparing against the content of the endpoint at `/promregatorMetrics`?

**Comment by eaglerainbow** [comment]
> Would it improve the situation for you, if it did (e.g. adding the docker container's ID as tag)?

Side note: https://github.com/mattsheppard/gethostname4j could be of help here to detect the container id, if we can detect that we are running in a container.
Alternatively, provide id parameter via Spring configuration externalization + setting the value via -Dxxx.xxx=`hostname` from `promregator.sh`. By this, users could also overwrite the value...

**Comment by vervas** [comment]
> Well, it kind of does 😁 with the endpoint of /promregatorMetrics (see also the endpoint page)

This part I completely missed 😇 I never filled the gap in my head, when moving from Single Endpoint Scraping to Single Target Scraping, and I was left wondering... (my belief was that it appended its metrics on every target's endpoint which would have been very strange). I should have read the docs ;) They have actually a very rich set of information would could be really useful.

Regarding the tagging, I think it would really help to have them adding a hostname or container ID, so that the different instances can be distinguished. An option to omit adding itself to the target from `/discovery` could help, with someone then explicitly adding all the promregator instances in his prometheus targets. That could be useful on any setup of promregator.

To a further extend, when running promregator on PCF, one could add promregator itself to the list of promregator targets. It is kind of a "meta" solution, since promregator would then proxy itself, which in turn could affect the results it returns for it.

**Comment by eaglerainbow** [comment]
> An option to omit adding itself to the target from `/discovery` could help, with someone then explicitly adding all the promregator instances in his prometheus targets.

What you suggest here implies that we would have to change the default behavior in Single Target Scraping mode: As you experienced, the `/discovery` endpoint contains the ´/promregatorMetrics` endpoint by default. I think this also very reasonable, as most consumers would expect to still get the promregator metrics when switching between the two modes (as you did and wondered how that worked  :grin:). Though, providing a configuration to disable that behavior (whose default value would be "add it to `discovery´), isn't a big thing and can be done easily. If you then "go HA setup", then you are free to do so. Yet, such a thing should be mentioned in something like a "HA setup page", as I assume many would forget to do it. 

> To a further extend, when running promregator on PCF, one could add promregator itself to the list of promregator targets.

Nice idea, but that would mean that Promregator "would need to know where it is located itself". That would be a very special configuration option, which 99% of non-users most likely would totally confuse ;-) Also testing such a thing could be hell...
But the idea has a lot of pcharm...

**Comment by vervas** [comment]
> Very charming side-effect of this: As the normal scraping procedure is happening, label enrichment on promregator's metrics is happening automatically, i.e. instanceId will be set accordingly for all promregator_* metrics, too....

Precisely :)

Regarding the exclusion of the exclusion of `/promregatorMetrics ` from the `/discovery` endpoint, one could simply do this outside of promregator with `jq` like `jq 'map(select(.labels.__meta_promregator_target_path != "/promregatorMetrics"))'`. That's a very easy solution without a great need to implement such a feature in promregator.

**Comment by eaglerainbow** [comment]
> That's a very easy solution without a great need to implement such a feature in promregator.

You are right. But implementing that feature roughly costs me a smile 😃  (and one little unit test). Then, hundred of consumers don't need to "fiddle around" with hack-ish `jq` statement (which might need to change, in case we will adjust the path name at some point in the future).

I'd like to propose a small deal here to you:
* I'll implement the necessary stuff (also the issue with the recursive request mentioned above needs to be addressed) and will tailor you a docker image suitable for testing.
* Can you then test our idea in your CF environment?

I'd also spend another markdown documentation page on the HA topic, giving guidance how such a setup could work. I'd greatly appreciate your input on that documentation!

And yes, for sure, this will only happen after you are back from your trip for sure!

**Comment by eaglerainbow** [comment]
I have created a `0.4.0-SNAPSHOT-hasetuptest1` for you - and it's uploaded to `hub.docker.com`. 

* It corresponds to branch `ha-setup-test` here. 
* The "HA setup" page is at https://github.com/promregator/promregator/blob/ha-setup-test/docs/ha-setup.md . I have marked the block still to be written with "to be provided with experience report". Please also feel free to change the entire part of the documentation (even the structure), if you have a better idea how to convey the message!
* The switch to toggle is called `promregator.discovery.ownMetricsEndpoint` (set it to `false`). Then the `/promregatorMetrics` endpoint will be gone in `/discovery`. 
* If you are able to do so, please also try to abuse it by forgetting to set `path: /promregatorMetrics`. be03a8131b9bd8dcacf9c2138cbe860618934cd6 should ensure that you then get a graceful error message and do not crash the machine with an endless loop.

If you find any bug, please let me know!

PS: Note that this version is based on 0.4.0 (that's `master`). Thus, it also contains a larger number of additional features, which aren't released yet, but are planned to be released. Therefore, it may be a bit unstable -- `0.4.0-SNAPSHOT-hasetuptest1` is purely meant for testing our HA idea here...


### DISCLAIMER to other readers
The test version `0.4.0-SNAPSHOT-hasetuptest1` provided via the Docker Hub is not meant for productive usage by you. Moreover, don't expect that the version will be usable for you and will stay for long on the Docker Hub! Once testing has been completed, the version will be removed again! Best simply ignore this version as a whole.

You have been warned!

**Comment by vervas** [comment]
> Then, hundred of consumers don't need to "fiddle around" with hack-ish jq statement (which might need to change, in case we will adjust the path name at some point in the future).

Completely agreed.

> * Can you then test our idea in your CF environment?
>
> I'd also spend another markdown documentation page on the HA topic, giving guidance how such a setup could work. I'd greatly appreciate your input on that documentation!

More than happy to do both :)

**Comment by vervas** [comment]
Unfortunately version `0.4.0-SNAPSHOT-hasetuptest1` seems to suffer from something similar to https://github.com/promregator/promregator/pull/55#issuecomment-388392384.

Not sure if something differs in the build but I am unable to test this. I gave it a few tries the past days with different configs but the result is the same. I will for now give a try to manually filter out the `/promregatorMetrics` target with version 0.3.6 and get back with results.

**Comment by eaglerainbow** [comment]
> Unfortunately version 0.4.0-SNAPSHOT-hasetuptest1 seems to suffer from something similar to #55 (comment).

That's not nice at all - as this means that the issue is not build-related: `0.4.0-SNAPSHOT-hasetuptest1` was built exactly the same way (same Jenkins, same server, same slave docker image,...) as I do create releases made public later on. 

> Not sure if something differs in the build but I am unable to test this. 

if you can make out some pattern of the problem, and you think it's related to Promregator, please open other issue here. This would be a major quality issue, which I would like to get fixed.

> I will for now give a try to manually filter out the /promregatorMetrics target with version 0.3.6 and get back with results.

Ok, sounds like a plan.

**Comment by vervas** [comment]
> That's not nice at all - as this means that the issue is not build-related

It actually looks like there is a bug in the code. When invoking the discovery function the app basically crashes with the same error as #59. This occurs on every request, so I believe something must be falling in an infinite loop or something. I tried this with multiple settings, like with and without  promregator in the targets list, and the toggle `promregator.discovery.ownMetricsEndpoint` set to both values.

This occurs consistently on `0.4.0` while `0.3.6` doesn't seem to be suffering from the issue in the first place.

> I will for now give a try to manually filter out the /promregatorMetrics target with version 0.3.6 and get back with results.

Looks like promregator is ignoring itself from the discovery targets 🤔. I will add some debug flags there to understand what is going on.

**Comment by eaglerainbow** [comment]
> When invoking the discovery function the app basically crashes with the same error as #59.

Mmhm... I can't reproduce this here, but all the cases of #59 are hardly reproducible here. 
That is why I have released `0.4.0-rc2` now, which contains the fixed timeout value (see also #59).

Are you able verify that your issue is also related to the communication to the CF cloud controller?

On `0.4.0-rc*` you might want to set
```yaml
logging:
  level:
    org:
      cloudfoundry:
         promregator: DEBUG
    reactor:
      mono:
        OnAssembly: INFO
      ipc:
        netty: DEBUG
promregator:
  reactor:
    debug: true
```
I have added a couple of additional logging points, which will be enabled by this.
*Warning!* This will (also) log the traffic between Promregator and the CF Cloud Controller! It may also contain your credentials to the API host and possibly also the access JWT. So, please don't make the logs public here without proper cleansing!

>  Looks like promregator is ignoring itself from the discovery targets

That should not be the case, fi have put `path: /promregatorMetrics` in the target. If not, then we scrape with the default `/metrics` and the new `isLoopbackRequest` comes into play (see https://github.com/promregator/promregator/blob/a84ad8ca6a5427a057aa518ca5a4caf87af4eb01/src/main/java/org/cloudfoundry/promregator/endpoint/AbstractMetricsEndpoint.java#L245 ). Remember though, that this check only exists in your special `0.4.0-SNAPSHOT-hasetuptest1` (on `0.4.0-rc*` that coding does not exist).

**Comment by vervas** [comment]
> Looks like promregator is ignoring itself from the discovery targets

The issue here was authentication (since I had `promregator.metrics.auth` set to BASIC). It might make sense to be able to make requests with basic auth against the defined targets in general but this is not related to this issue. I will have a look in the code to see if I can implement something like that (it has been a long while since I last wrote Java code).

**Comment by eaglerainbow** [comment]
Oh, that's an interesting coincidence: We are running now into a "mixed-authenticator scenario" here. Eloo already created an issue (#60) for that case. 

Before you deep-dive on the "authenticator per targets", can you confirm that at least *without* authentication it works?

**Comment by eaglerainbow** [comment]
Just for the record: I just stumbled over a thing, which we should document on the HA page: on https://github.com/promregator/promregator/blob/f4699c885ddb1db76c59f0f312e5b72e914ad87e/docs/config.md#option-promregatorauthenticationbasicpassword-optional it states that

> If not specified, a random password is generated during startup. Its value is printed to the standard error device.

This wouldn't work in a HA setup: The generated password would be different for each instance as it is taken randomly. So, the consumer wouldn't know which password to write down in the "authenticator" section. The best would be to have the password set in the config file explicitly.

**Comment by vervas** [comment]
> Before you deep-dive on the "authenticator per targets", can you confirm that at least without authentication it works

Yeah I have the promregator properly scraping itself.

> The best would be to have the password set in the config file explicitly

I think this should be the case anyhow; I cannot think of any use cases where one generates a random password on startup and then prints it out, but I might be wrong.

Besides these, atm my only issue is that the `promregator.discovery.ownMetricsEndpoint` seems to have no effect but I'm trying to figure out what is going on there.

**Comment by eaglerainbow** [comment]
> Besides these, atm my only issue is that the promregator.discovery.ownMetricsEndpoint seems to have no effect but I'm trying to figure out what is going on there.

Ok, great. Please feel free to call out in case you need assistance.

**Comment by eaglerainbow** [comment]
I just created&uploaded you a version `0.4.0-SNAPSHOT-hasetuptest2` which is based on the state as of `0.4.0-rc3` (+ some spring boot version bump).  This is to give you the chance to test the HA stuff with the most recent state of the code on `master` (NB: `0.4.0-SNAPSHOT-hasetuptest1` still is on state as of pre-`0.4.0-rc1`).

Yet, I am a little bit reluctant to merge the branch "ha-setup-test" into "master" due to your statement

> Besides these, atm my only issue is that the promregator.discovery.ownMetricsEndpoint seems to have no effect but I'm trying to figure out what is going on there.

**Comment by eaglerainbow** [comment]
As you were! There is build error for `0.4.0-SNAPSHOT-hasetuptest2`; I have to check it and will report back...

**Comment by eaglerainbow** [comment]
Finally got it working (as usual in the end it had only been a small glitch): `0.4.0-SNAPSHOT-hasetuptest2` is there to be tested now...

**Comment by vervas** [comment]
I can confirm that `0.4.0-SNAPSHOT-hasetuptest2` works correctly! I'll run this for a while and share my experience with it.

**Comment by eaglerainbow** [comment]
Sounds good.
I'd suggest that I'll do #60 for you first and then build you a `hasetuptest3` which also contains the authenticator-stuff. Then you can do a longer full-fledged test before we merge it completely to `master`

**Comment by eaglerainbow** [comment]
@vervas : Here's a warning for you! 
`0.4.0-SNAPSHOT-hasetuptest2` is also affected of the severe path-random bug mentioned in #59 likewise `0.4.0-rc3` and earlier. You might go crazy with it. 
It will only be fixed with the next `-hasetuptest` which I will create for you.

**Comment by eaglerainbow** [comment]
There is a `0.4.0-SNAPSHOT-hasetuptest3` for you now, which also contains the implementation as mentioned in #60. 
If you want, you may give it all a try together...

**Comment by vervas** [comment]
Running `0.4.0-SNAPSHOT-hasetuptest3` on CF with basic auth enabled on metricsworks without an issue.

Running locally against the same CF targets somehow produces
https://github.com/promregator/promregator/blob/aaedc7ed0eb8ba35d1beb25b343b21f01c2ff857/src/main/java/org/cloudfoundry/promregator/config/validations/TargetsHaveConsistentAuthenticatorId.java#L34-L36.
It is a bit strange but the actual deployment works fine.

**Comment by eaglerainbow** [comment]
> Running locally against the same CF targets somehow produces [validation error message]

And, is the error message correct? Does the target refer to an authenticator, which is not defined in the configuration?

The unit test class `TargetsHaveConsistentAuthenticatorIdTest` runs just fine for me (and contains both positive and negative cases). Perhaps you have found a new special case, which is not covered properly, yet?
Are you able to post an excerpt of your configuration here?

**Comment by vervas** [comment]
My initial statement was inaccurate. The error occurs both locally and on CF. Here is my configuration which looks good to me but I might be missing something:
```
promregator:
  authenticator:
    type: none

  authentication:
    basic:
      username: USERNAME
      password: PASSWORD

  targetAuthenticator:
    - id: promregatorAuth
      type: basic
      basic:
        username: USERNAME
        password: PASSWORD

  endpoint:
    auth: BASIC
  discovery:
    hostname: HOSTNAME
    port: 443
    ownMetricsEndpoint: false
    auth: BASIC
  metrics:
    auth: BASIC

  targets:
    ...
    - orgName: ORG_NAME
      spaceName: SPACE_NAME
      applicationName: promregator
      path: "/promregatorMetrics"
      authenticatorId: promregatorAuth
```

**Comment by eaglerainbow** [comment]
I was able to reproduce it here locally - I'll have a look from where the problem is coming. Afterwards, I'll come back to you.

**Comment by eaglerainbow** [comment]
😀 
Well, the solution to this issue is too trivial:

In your configuration file, could you please rename your block

```yaml
promregator:
  targetAuthenticator:
```

to read

```yaml
promregator:
  targetAuthenticators:
```
(with an `s` at the end)?

NB: The documentation was wrong on this and I have fixed the latter with 0fdb4d26b526da1acbe0becad59af49fcc2b81c8.

**Comment by vervas** [comment]
Yeap that was too obvious to spot :D 

I can confirm this works perfect now. Both the authenticator and authenticating and pulling logs out of multiple instances or promregator.

That said I consider this issue resolved. I will provide some docs about it and you can close it either before or after.

Thanks a lot for all the support again!

**Comment by eaglerainbow** [comment]
Thank YOU for the PR - and all the testing!

**Comment by eaglerainbow** [comment]
Just for the record: I have cleaned up the images `0.4.0-SNAPSHOT-hasetuptest*` again. 
Branch `ha-setup` has been merged into `master`, too, so the result of this HA setup test will be included with the 0.4.0 release.
